### ISSUE TITLE
For some inputs inner_join(..., na_matches = "never") returns rows with NA in the matched column.

### ISSUE THREAD

**Issue by antalk2** [tidyverse/dplyr#5568]
Problem: for some inputs `inner_join(..., na_matches = "never")` returns
rows with `NA` in the matched column.

<details>

## Introduction

`inner_join` has a parameter `na_matches`. 
According to its documentation:

>  Use "never" to always treat two NA or NaN values as different

## Load libraries

```r
library(devtools)
```

```
## Loading required package: usethis
```

```r
install_github("tidyverse/dplyr")
```

```
## Skipping install of 'dplyr' from a github remote, the SHA1 (dffe1727) has not changed since last install.
##   Use `force = TRUE` to force installation
```

```r
library(dplyr)
```

```
## 
## Attaching package: 'dplyr'
```

```
## The following objects are masked from 'package:stats':
## 
##     filter, lag
```

```
## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union
```

```r
library(tibble)
```

## Code working as expected

- `inner_join()` yields row `(NA,NA,NA)`
- `inner_join(..., na_matches = "never")` omits rows with NA in the matched
column.

```r
x = tribble(
   ~xid,   ~xval,
   NA      , NA,
   'a'     , NA,
   'b'     , 'a'
)
y = tribble(
   ~yid,   ~yval,
   NA      , NA,
   'a'     , NA,
   'b'     , 'a'
)
inner_join( x, y, by=c('xval'='yid') )
```

```
## # A tibble: 3 x 3
##   xid   xval  yval 
##   <chr> <chr> <chr>
## 1 <NA>  <NA>  <NA> 
## 2 a     <NA>  <NA> 
## 3 b     a     <NA>
```

```r
inner_join( x, y, by=c('xval'='yid'), na_matches = "never" )
```

```
## # A tibble: 1 x 3
##   xid   xval  yval 
##   <chr> <chr> <chr>
## 1 b     a     <NA>
```

## However the following fails

- `inner_join()` yields row `(NA,NA,NA)` as expected
- `inner_join(..., na_matches = "never")` does NOT omit rows
with NA in the matched column.

```r
x = tribble(
   ~xid,    ~xval,
   NA       , NA,
   '1.Béla' , NA,
   '2.Béla' , '1.Béla'
)
y = tribble(
   ~yid,    ~yval,
   NA       , NA,
   '1.Béla' , NA,
   '2.Béla' , '1.Béla'
)
inner_join( x, y, by=c('xval'='yid') )
```

```
## # A tibble: 3 x 3
##   xid    xval   yval 
##   <chr>  <chr>  <chr>
## 1 <NA>   <NA>   <NA> 
## 2 1.Béla <NA>   <NA> 
## 3 2.Béla 1.Béla <NA>
```

```r
inner_join( x, y, by=c('xval'='yid'), na_matches = "never" )
```

```
## # A tibble: 3 x 3
##   xid    xval   yval 
##   <chr>  <chr>  <chr>
## 1 <NA>   <NA>   <NA> 
## 2 1.Béla <NA>   <NA> 
## 3 2.Béla 1.Béla <NA>
```

## Test

```r
library(testthat)
test_that("inner_join omits NA from matched column", {
x = tribble(
   ~xid,    ~xval,
   NA       , NA,
   '1.Béla' , NA,
   '2.Béla' , '1.Béla'
)
y = tribble(
   ~yid,    ~yval,
   NA       , NA,
   '1.Béla' , NA,
   '2.Béla' , '1.Béla'
)
out <- inner_join( x, y, by=c('xval'='yid'), na_matches = "never" )
  expect_equal(any(is.na(out$xval)), FALSE)
})
```

## SessionInfo

```r
sessionInfo()
```

```
## R version 3.6.3 (2020-02-29)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.5 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=hu_HU.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=hu_HU.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=hu_HU.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=hu_HU.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] tibble_3.0.4     dplyr_1.0.2.9000 devtools_2.3.0   usethis_1.6.1   
## 
## loaded via a namespace (and not attached):
##  [1] pillar_1.4.6      compiler_3.6.3    prettyunits_1.1.1 remotes_2.1.1    
##  [5] tools_3.6.3       testthat_2.3.2    digest_0.6.26     pkgbuild_1.0.8   
##  [9] pkgload_1.1.0     lifecycle_0.2.0   evaluate_0.14     memoise_1.1.0    
## [13] pkgconfig_2.0.3   rlang_0.4.8       cli_2.1.0         curl_4.3         
## [17] yaml_2.2.1        xfun_0.15         withr_2.2.0       stringr_1.4.0    
## [21] knitr_1.29        desc_1.2.0        generics_0.0.2    fs_1.4.2         
## [25] vctrs_0.3.4       tidyselect_1.1.0  rprojroot_1.3-2   glue_1.4.2       
## [29] R6_2.4.1          processx_3.4.2    fansi_0.4.1       rmarkdown_2.4    
## [33] sessioninfo_1.1.1 purrr_0.3.4       callr_3.4.3       magrittr_1.5     
## [37] backports_1.1.8   ps_1.3.3          ellipsis_0.3.1    htmltools_0.5.0  
## [41] assertthat_0.2.1  utf8_1.1.4        stringi_1.4.6     crayon_1.3.4
```

</details>

**Comment by hadley** [comment]
Can you please make your reprex much much more concise? I suspect you can explain the problem in two paragraphs instead of two pages, and that's much faster for me to process. Also, as a general rule, there's no need to provide `sessionInfo()` unless it's explicitly asked for; it doesn't usually provide any additional insight and just clutters up the issue further.

**Comment by antalk2** [comment]
## Problem

Recoding values changes behaviour of
`inner_join( z, z, by=c('xval'='xid'), na_matches = "never" )`


```r
library(dplyr)
library(tibble)
x = tribble(
   ~xid,   ~xval,
   NA      , NA,
   'a'     , NA,
   'b'     , 'a'
)
# y is the same as x, except uses different labels
MAP = c(a='1.Béla', b='2.Béla')
y = mutate(x, xid = MAP[xid], xval=MAP[xval] )

# this is what we do with both
f <- function(z){
  inner_join( z, z, by=c('xval'='xid'), na_matches = "never" )
}

xout <- f(x)
yout <- f(y)
```


One would expect `yout` and `xout` to be identical,
except for different labels for the same values.
But `yout` contains extra rows, with `NA` in the `xval` column.


```r
xout
```

```
## # A tibble: 1 x 3
##   xid   xval  xval.y
##   <chr> <chr> <chr> 
## 1 b     a     <NA>
```

```r
yout
```

```
## # A tibble: 3 x 3
##   xid    xval   xval.y
##   <chr>  <chr>  <chr> 
## 1 <NA>   <NA>   <NA>  
## 2 1.Béla <NA>   <NA>  
## 3 2.Béla 1.Béla <NA>
```

**Comment by hadley** [comment]
Somewhat more minimal reprex:

``` r
library(dplyr, warn.conflicts = FALSE)

df1 <- tribble(
   ~x,  ~y,
   NA,    NA,
   'a',   NA,
   'b',   'a'
)
df2 <- df1 %>% select(x = y, z = x)

inner_join(df1, df2, by = "x")
#> # A tibble: 3 x 3
#>   x     y     z    
#>   <chr> <chr> <chr>
#> 1 <NA>  <NA>  <NA> 
#> 2 <NA>  <NA>  a    
#> 3 a     <NA>  b
inner_join(df1, df2, na_matches = "never", by = "x")
#> # A tibble: 1 x 3
#>   x     y     z    
#>   <chr> <chr> <chr>
#> 1 a     <NA>  b


df3 <- tribble(
   ~x, ~y,
   NA,   NA,
   'Aé', NA,
   'Bé', 'Aé'
)
df4 <- df3 %>% select(x = y, z = x)

inner_join(df3, df4, by = "x")
#> # A tibble: 3 x 3
#>   x     y     z    
#>   <chr> <chr> <chr>
#> 1 <NA>  <NA>  <NA> 
#> 2 <NA>  <NA>  Aé   
#> 3 Aé    <NA>  Bé
inner_join(df3, df4, na_matches = "never", by = "x")
#> # A tibble: 3 x 3
#>   x     y     z    
#>   <chr> <chr> <chr>
#> 1 <NA>  <NA>  <NA> 
#> 2 <NA>  <NA>  Aé   
#> 3 Aé    <NA>  Bé
```

<sup>Created on 2020-11-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by hadley** [comment]
Slightly simpler reprex:

``` r
library(dplyr, warn.conflicts = FALSE)

df1 <- tibble(x = c(NA, "e"))
df2 <- tibble(x = c(NA, "e"), y = seq_along(x))
inner_join(df1, df2, by = "x")
#> # A tibble: 2 x 2
#>   x         y
#>   <chr> <int>
#> 1 <NA>      1
#> 2 e         2
inner_join(df1, df2, na_matches = "never", by = "x")
#> # A tibble: 1 x 2
#>   x         y
#>   <chr> <int>
#> 1 e         2

df1 <- tibble(x = c(NA, "é"))
df2 <- tibble(x = c(NA, "é"), y = seq_along(x))
inner_join(df1, df2, by = "x")
#> # A tibble: 2 x 2
#>   x         y
#>   <chr> <int>
#> 1 <NA>      1
#> 2 é         2
inner_join(df1, df2, na_matches = "never", by = "x")
#> # A tibble: 2 x 2
#>   x         y
#>   <chr> <int>
#> 1 <NA>      1
#> 2 é         2
```

<sup>Created on 2020-11-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by hadley** [comment]
Looks like a vctrs issue — I've filed an issue there.

**Comment by DavisVaughan** [comment]
Fixed with CRAN vctrs 0.3.5

``` r
library(dplyr, warn.conflicts = FALSE)
#> Warning: package 'dplyr' was built under R version 4.0.2

df1 <- tibble(x = c(NA, "e"))
df2 <- tibble(x = c(NA, "e"), y = seq_along(x))
inner_join(df1, df2, by = "x")
#> # A tibble: 2 x 2
#>   x         y
#>   <chr> <int>
#> 1 <NA>      1
#> 2 e         2
inner_join(df1, df2, na_matches = "never", by = "x")
#> # A tibble: 1 x 2
#>   x         y
#>   <chr> <int>
#> 1 e         2

df1 <- tibble(x = c(NA, "é"))
df2 <- tibble(x = c(NA, "é"), y = seq_along(x))
inner_join(df1, df2, by = "x")
#> # A tibble: 2 x 2
#>   x         y
#>   <chr> <int>
#> 1 <NA>      1
#> 2 é         2
inner_join(df1, df2, na_matches = "never", by = "x")
#> # A tibble: 1 x 2
#>   x         y
#>   <chr> <int>
#> 1 é         2
```

<sup>Created on 2020-11-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by hadley** [comment]
We should bump the vctrs requirement to close this issue.
